### PR TITLE
Search unification

### PR DESCRIPTION
**Pull request by epsilon-phase**
part of solving #18875 

I've attempted to follow @coolthulu's advice about writing a unified search function that can be used to replace the many different item searches that exist throughout the game.

It isn't stable yet, and I'm not sure that I've done this in an acceptable way, so criticism is welcome.

### CODE REVIEW

File: src/item_search.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#include "item_search.h"
+#include "cata_utility.h"
+
+std::pair<std::string, std::string> get_both( const std::string &a );
+std::function<bool( const item & )>
+item_filter_from_string(  std::string filter )
+{
+    if( filter.empty() ){
+        return [](const item&i){
+            return i.active || !i.active;
```
Comment (Coolthulhu) [inline src/item_search.cpp]: What?

**Comment by Coolthulhu** [comment]
Needs cleanup, but otherwise it looks good.

File: src/item_search.cpp (hunk 1 above)
Comment (jaked122) [inline src/item_search.cpp]: It requires a true value, and the compiler flags don't accept not using the parameter. :P

File: src/item_search.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            };
+            break;
+        case 'e'://either
+            return [filter]( const item & i ) {
+                auto pair = get_both( filter );
+                return item_filter_from_string( pair.first )( i )
+                       || item_filter_from_string( pair.second )( i );
+            };
+            break;
+        case 'b'://both
```
Comment (jaked122) [inline src/item_search.cpp]: This is different, this allows searching by both filters at once, rather than one or the other.

**Comment by jaked122** [comment]
So now the inventory search actually exists, courtesy of an extra vector of indexes of items that match the query.

File: src/item_search.cpp (hunk 2 above)
Comment (jaked122) [inline src/item_search.cpp]: @Coolthulhu, now the search has the ability to use the comma instead.

**Pull request merged**